apps/shared/avifpng.c

```diff
@@ -104,61 +108,132 @@ static avifBool avifCopyRawProfile(const char * profile, size_t profileLength, a
-        if ((exifSize == 0) || !exif) {
-            fprintf(stderr, "Exif extraction failed: empty eXIf chunk\n");
-            return AVIF_FALSE;
+    if (!*ignoreExif) {
+        png_uint_32 exifSize = 0;
+        png_bytep exif = NULL;
+        if (png_get_eXIf_1(png, info, &exifSize, &exif) == PNG_INFO_eXIf) {
+            if ((exifSize == 0) || !exif) {
+                fprintf(stderr, "Exif extraction failed: empty eXIf chunk\n");
+                return AVIF_FALSE;
+            }
+            avifImageSetMetadataExif(avif, exif, exifSize);
+            *ignoreExif = AVIF_TRUE; // Ignore any other Exif chunk.
         }
-        avifImageSetMetadataExif(avif, exif, exifSize);
-        return AVIF_TRUE;
     }
 #endif // PNG_eXIf_SUPPORTED
 
+    // tXMP could be retrieved using the png_get_unknown_chunks() API but tXMP is deprecated
+    // and there is no PNG file example with a tXMP chunk lying around, so it is not worth the hassle.
+
     png_textp text = NULL;
     const png_uint_32 numTextChunks = png_get_text(png, info, &text, NULL);
-    for (png_uint_32 i = 0; i < numTextChunks; ++i, ++text) {
-        if (!strcmp(text->key, "Raw profile type exif") || !strcmp(text->key, "Raw profile type APP1")) {
-            png_size_t textLength;
-            switch (text->compression) {
-#ifdef PNG_iTXt_SUPPORTED
-                case PNG_ITXT_COMPRESSION_NONE:
-                case PNG_ITXT_COMPRESSION_zTXt:
-                    textLength = text->itxt_length;
-                    break;
-#endif
-                case PNG_TEXT_COMPRESSION_NONE:
-                case PNG_TEXT_COMPRESSION_zTXt:
-                default:
-                    textLength = text->text_length;
-                    break;
+    for (png_uint_32 i = 0; (!*ignoreExif || !*ignoreXMP) && (i < numTextChunks); ++i, ++text) {
+        avifRWData metadata = { NULL, 0 };
+        avifBool tagExif00WasRemoved;
+        if (!*ignoreExif && !strcmp(text->key, "Raw profile type exif")) {
+            if (!avifCopyRawProfile(text->text, avifGetPngTextLength(text), &metadata, &tagExif00WasRemoved)) {
+                return AVIF_FALSE;
+            }
+            uint32_t exifTiffHeaderOffset;
+            if (avifExtractExifTiffHeaderOffset(metadata.data, metadata.size, &exifTiffHeaderOffset) != AVIF_RESULT_OK) {
+                avifRWDataFree(&metadata);
+                fprintf(stderr, "Exif extraction failed: missing TIFF header\n");
+                return AVIF_FALSE;
+            }
+            avifRWDataFree(&avif->exif);
+            avif->exif = metadata;
+            *ignoreExif = AVIF_TRUE; // Ignore any other Exif chunk.
+        } else if (!*ignoreXMP && !strcmp(text->key, "Raw profile type xmp")) {
+            if (!avifCopyRawProfile(text->text, avifGetPngTextLength(text), &metadata, &tagExif00WasRemoved)) {
+                return AVIF_FALSE;
+            }
+            if (tagExif00WasRemoved) {
```

Comment:
Don't think this'll ever happen, but why not. Would've been better to check if there is a valid XML packet somehow (like checking there is valid Exif above), but probably an overkill...

Reply:
Is there an official list of valid XMP headers? From crawled JPEGs I saw mostly `<?xpacket` but I found a few `<x:xmpmeta`, `?xpacket`, `http://ns.adobe.com/xap/1.0/`, `-<rdf:RDF`, etc.

Also in libwebp the spirit was: do not care about metadata content, just pass it around. I do not know to what extend we want to apply that in libavif.

Reply:
IMHO anything that starts w/ a "<" should be a bare (valid) XML load with no header that should be kept (i.e. as coming from TIFFs, Jpeg2000, compliant PNGs etc.). The only JPEG APP1 header I'm aware of is the `http://ns.adobe.com/xap/1.0/\0` - see the [XMP spec for JPEG](https://github.com/adobe/xmp-docs/blob/master/XMPSpecifications/XMPSpecificationPart3.pdf); [ImageMagick also implements this for JPEG](https://github.com/ImageMagick/ImageMagick/blob/80bd592dd20d4a1287842459ca3801c3ae3866cd/coders/jpeg.c#L185), and I think it is safe to assume this in the PNG case for "Raw profile type APP1" (and most probably also for "Raw profile type xml"), as, like "Raw profile type exif", these seem to be verbatim copies of the corresponding JPEG APP1 payloads before the eXIF chunk and "XML:com.adobe.xmp" keys were specified for the headerless payloads.

Caveat: note that there is also the "ExtendedXMP" JPEG APP1 header `http://ns.adobe.com/xmp/extension/\0` that is used for concatenated packets to work around the 64kB JPEG APP1 limit, so supporting fusion of multiple APP1 segments when reading JPEGs might be required as well (already done by e.g. exiv2 and XMPSdk). Be aware that there is also a possibility of concatenated Exif APP1 segments in JPEG as well.

Reply:
> I now realize that it is the same behavior between PNG raw profiles and Jpeg APP1, for the header part.

Great! I think for "Raw profile type APP1" it must be the same (what other choice is there?); for "Raw profile type exif" and "Raw profile type xmp" it should be the same (as e.g. implemented by ImageMagick), but again, as there is no official spec, there are no guarantees so better check for headers just in case, but one should also allow the case when they're not there as the key kinda identifies the data...